Adds serial port detection for Z Wave USB Z Sticks

kdschlosser opened the pull request:
## Description:
adds serial port detection for the USB Z Wave sticks

I currently only have the Vendor ID for the Aeon Labs Gen5 Z Stick. others will need to be added for the detection to work for those brands.

There is a constant called VENDOR_IDS in homeassistant/components/zwave/config_flow.py this is where other Z Sticks would need to be added.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

kdschlosser committed:
adds pyserial to reequirements

kdschlosser committed:
adds com port detection

I am using pyserial to enumerate the ports on the local computer. this enumeration yields a serial port information object which supplies the Vendor ID (VID) of the device that created the port. There is a list of vendor name, VID that is enumerated to ses if the VID in the list matches the one for the port. if it does it then returns the name and port. this gets placed into the data flow.

according to the documentation for pyserial this mechanism should work for MAC OSX, Linux, and Windows and possibly other OS's. The documentation for pyserial names those 3 specifically and states that the information "may" not be there for other OS's
I can confirm it does work on Windows.

kdschlosser committed:
Adds missing sys import

kdschlosser committed:
Removes white space

kdschlosser committed:
Fixes line length being greater then 79

fabaff commented on homeassistant/components/zwave/config_flow.py:
```diff
@@ -31,9 +48,13 @@ def __init__(self):
 
         errors = {}
 
+        com = _get_z_stick()
```
Please rename the variable to something meaningful.

fabaff commented on homeassistant/components/zwave/config_flow.py:
```diff
@@ -12,6 +13,22 @@
+VENDOR_IDS = (('Aeon Labs Gen5', 'VID_0658'),)
+
+
+def _get_z_stick():
```
Docstring is missing.

fabaff commented on homeassistant/components/zwave/config_flow.py:
```diff
@@ -12,6 +13,22 @@
+        if port.vid is None:
+            continue
+        for vendor, vid in VENDOR_IDS:
+            if vid == 'VID_' + hex(port.vid)[2:].zfill(4):
```
Please use string formatting.

fabaff commented on homeassistant/components/zwave/__init__.py:
```diff
@@ -42,7 +42,11 @@
+REQUIREMENTS = [
+    'pydispatcher==2.0.5', 
+    'homeassistant-pyozw==0.1.2', 
+    'pyserial==3.4'
```
All integrations must use the same version.

Also, new or updated dependencies must be added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

fabaff reviewed (changes requested):
Perhaps we should think about adding this kind of feature to `netdisco`. There are a bunch of integrations which are dealing with USB-connected devices.

kpine commented on homeassistant/components/zwave/config_flow.py:
```diff
@@ -12,6 +13,22 @@
+                if sys.platform.startswith('win'):
+                    return vendor + ' : ' + '\\\\.\\' + port.device
+                else:
+                    return vendor + ' : ' + port.device
```
So, I think there's a problem with this approach of including the vendor name. For an Aeotec stick (using Linux paths as an example), this might result in something like `Aeon Labs Gen5 : /dev/ttyACM0`, right?

Well, the VID 0x0658 refers to Sigma Labs, not Aeon Labs/Aeotec. Other USB sticks that use the Sigma Labs (now Silicon Labs) Z-Wave SOC report the same VID. Besides the Aeotec I also have a Silicon Labs UZB3 that reports 0658. I think the UZB1 also has the same VID. This is going to be confusing for people not using the Aeotec Z-Stick. I don't really know how many USB sticks out there use the Sigma design, but it's at least 3.

If you wanted to also use the PID (product id), you can't really. The Aeotec reports a PID of 0x200. So does the UZB3. Who knows how many other sticks out there also report 0658:0200.

In the end, it's not going to be possible to accurately describe every USB stick by name, and calling all SIgma-based sticks "Aeon Labs/Aeotec" is not right. I would recommend just annotating the path name with some generic indicator, instead of a device name, e.g. `Z-Wave USB : <usb_path>` or something, or just leave off the description entirely and still provide the suggestion.

kpine commented on homeassistant/components/zwave/config_flow.py:
```diff
@@ -12,6 +13,22 @@
+    for port in serial.tools.list_ports.comports(include_links=False):
+        if port.vid is None:
+            continue
+        for vendor, vid in VENDOR_IDS:
```
So if you were looking to support other USB sticks, there might be some special exceptions to this simple logic, for example the HUSBZB-1 Z-Wave/Zigbee combo USB stick. It has a path for both the Z-Wave and Zigbee devices, and they both share the same VID, so which one would be chosen? You might return the Zigbee device path. Not a problem now, but more of an FYI.

On the plus side, I can confirm this does work well on Mac OS. Pyserial only returns a `/dev/cu.usbmodemFA131` path for my UZB3 stick. The alternative path`/dev/tty.usbmodemFA131` device, won't work and isn't listed.

kpine reviewed (commented):
This is a good idea, but I think there are a few problems with how it's implemented. See the code comments.

On Linux, it'd be a much better idea to recommend the persistent `/dev/serial/by-id` device paths instead. The Aoetec might be `/dev/ttyACM0`, or it could be `/dev/ttyACM1`. That path is not persistent and can change for a variety of reasons. When it does change, a user's Z-Wave configuration is completely busted, and there's no clue why. The usb path is also not editable once it's been configuraed, and `configuration.yaml` changes are ignored (that one alone confuses most). This would all be avoided by using the persistent symlinks instead, which will never change (aside from replacing the stick with a new brand!). If instead, the config flow recommended `/dev/serial/by-id/usb-0658_0200-if00` by default for the Aotec (assuming the symlink exists), that would produce a configuration that always has a valid usb path (again, assuming the stick is there).

I don't think you should implement the above in this change, I just though I'd point it out if anyone else is paying attention and is looking for ways to further improve the HA Z-Wave experience.

Gamester17 commented:
FYI, keep in mind that "Gen7" stick for Z-Wave 700 series is expected this year (2019). Z-Wave 700 platform was announced a while back but transceiver chips were first made available in December 2018

https://news.silabs.com/2018-12-12-Next-Generation-Z-Wave-700-Launches-on-the-Silicon-Labs-Wireless-Gecko-Platform

https://www.silabs.com/products/wireless/mesh-networking/z-wave/700-platform/what-is-z-wave-700

https://www.silabs.com/products/wireless/mesh-networking/z-wave/700-platform

kdschlosser commented:
I did not know all that information about the VID as I had hunted about trying to gather a list of them.. I came up empty handed. I simply put in the device I had as well as the VID for it. I can work around that. I did want to provide the user with the specific brand of the stick. so they would not think that the supplied serial port was by example. I am fairly confident i can get around the zigbee issue as well. 

let me work up a revision of this and see what you guys think.

It is a good starting point and it does work. I know this was one of the things I had read on the TODO list for Home Assistant and I had already done it so I knew how to go about it.

kdschlosser commented:
This design is also done so the user can still enter a port. so if there is not a matching VID then it will post the original example.

kpine commented:
> This design is also done so the user can still enter a port. so if there is not a matching VID then it will post the original example.

You're right, I was looking at the value passed to the ZWaveOption instance, not the part where the user input is saved to the config entry. Wouldn't you still want to pass the user input to that?

kpine commented:
So this change is not working for me. It finds my stick, but If I leave the form as the default `Aeon Labs Gen5 : /dev/ttyACM0`, I get an exception:

```
2019-02-18 18:11:17 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Z-Wave for zwave
Traceback (most recent call last):
  File "/home/keith/src/home-assistant/lib/python3.6/site-packages/openzwave/option.py", line 78, in __init__
    raise ZWaveException(u"Can't find device %s : %s" % (device, traceback.format_exception(*sys.exc_info())))
openzwave.object.ZWaveException: "Zwave Generic Exception : Can't find device Aeon Labs Gen5 : /dev/ttyACM0 : ['NoneType: None\\n']"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/keith/src/home-assistant/homeassistant/config_entries.py", line 258, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/home/keith/src/home-assistant/homeassistant/components/zwave/__init__.py", line 287, in async_setup_entry
    config_path=config.get(CONF_CONFIG_PATH))
  File "/home/keith/src/home-assistant/lib/python3.6/site-packages/openzwave/option.py", line 81, in __init__
    raise ZWaveException(u"Error when retrieving device %s : %s" % (device, traceback.format_exception(*sys.exc_info())))
openzwave.object.ZWaveException: 'Zwave Generic Exception : Error when retrieving device Aeon Labs Gen5 : /dev/ttyACM0 : [\'Traceback (most recent call last):\\n\', \'  File "/home/keith/src/home-assistant/lib/python3.6/site-packages/openzwave/option.py", line 78, in __init__\\n    raise ZWaveException(u"Can\\\'t find device %s : %s" % (device, traceback.format_exception(*sys.exc_info())))\\n\', \'openzwave.object.ZWaveException: "Zwave Generic Exception : Can\\\'t find device Aeon Labs Gen5 : /dev/ttyACM0 : [\\\'NoneType: None\\\\\\\\n\\\']"\\n\']'
```

This is because the `user_input` value is not being sanitized, instead the `com` variable is. `Aeon Labs Gen5 : /dev/ttyACM0` is being stored directly into the config entry.

kdschlosser committed:
Seveeral updates

Removes the vendor name from the pre defined list of vendors.
Removes VID_ from the vendor id in the pre defined list of vendors
Adds grabbing the vendor name from pyserial
Adds checking the description, interface name and the product name for the word Zigbee. this is to satisfy not grabbing the Z Sticks that support both Z Wave and Zigbee on a single USB stick.

kdschlosser commented:
if you want to have a go over the new commit i did. this removes most of what has been discussed. there are still several things that need to be changed. but I think for the most part this should fix that traceback and also supply the user with the correct port. 

I also added some checking to see if Zigbee is mentioned in the product name, the description and the interface name, if it is we skip over it. I am going to make an assumption that because the HUSBZB-1 supports both Z-Wave and Zigbee the interface name is going to be set to identify what port is what.

I could also add another check to make sure that UZB is in the description. I do not know if all of the manufacturers have UZB in the description. this may only be something that was done by Sigma Labs.

kdschlosser commented:
and as far as the symlinks are concerned. we can use that if needed there is also a serial number we can use from the stick.

Windows has a very similar issue where it will install a new set of drivers by simply plugging the stick into a different usb port. this is going to produce a different com port.

So instead of using the serial port as the qualifier. perhaps we should use the serial number. this can be done even if the user hand keys in a serial port. we just iterate over the ports find the matching one and then grab the serial number from the stick.

I do not know if this would be the way to go about it. the user config directory for python_openzwave would have to be names the same as the serial number for the stick for easy identification.

we can also then check to see if a stick is already added to hass. if it has been then we could continue the iteration over the ports looking for a possible second stick. and if there is one then supply that. and if there is not pop up a notification that on could not be located and then give the user the ability to enter the information.
